### PR TITLE
Developer msft

### PR DESCRIPTION
**Pull request by MeritedHobbit**
Regression Suite 0
image

Framework Changes
1. Execution mechanism using tags using include tags, exclude tags, test case tags and test group tags. 
2. Test case list (-l) now displayed in tabular form with tag information per test case and group. 
3. Multiple regression suites to choose from. (refer test_runner.json file from json_spec) 
4. dut_info json file includes IdleWaitTimeAfterFirmwareUpdate (default is 5mins to overcome issues related to rate limiting)
5. Now all test cases get invoked if option is provided (with only workspace). 

Test Case changes 
1. All test cases have been tagged with L0, L1, L2 , L3. L3 test cases are not a part of compliance. 
2. Removed unnecessary power cycles after many negative test cases. Only few test cases have this today. 
3. Service validator bugs fixed. Plus we have a progress bar too. 
4. Stagingtime and Activationtime are recorded for test run optimization (for F63 and F64). 
5. AC cycle test case (L3 test case) will not check for GPU goodness on return - reduces dependency during tool enabling. 
7. Reduced initial collateral dependency for ease of onboarding. 
8. Clear Log Dump test case is tolerant to redfish return code of 204 or 200 (despite no json body return in either case). 
9. Corrected file names of F24 (full device) and F62 (single device) negative test cases with AC reset interruption.

### CODE REVIEW

**Comment by v-ajitpanda** [comment]
> * Tested the current version for T0, R1, R2 and R3, and following are the issues found.
> 
> --	Issue	Category	Details	Resolution/ Current WAR
> 1	Bug in RAS test R3	Coding	"testRunArtifact": {         "log": {             "severity": "FATAL",             "message": "Traceback (most recent call last):\n  File "/home/afsanac/GitRepos/ocp-diag-ctam/ctam/test_runner.py", line 480, in _run_group_test_cases\n    test_result = test_instance.run()\n  File "/home/afsanac/GitRepos/ocp-diag-ctam/ctam/tests/ras/ctam_test_collect_crashdump_oem_selftest.py", line 71, in run\n    self.group.health_check_ifc.ctam_getes()\nAttributeError: 'BasicRasTestGroup' object has no attribute 'health_check_ifc'\n",             "sourceLocation": {                 "file": "/home/afsanac/GitRepos/ocp-diag-ctam/ctam/test_runner.py",                 "line": 487             }         }     },	 
> 2	T0: RF Service Validator	Coding	subprocess.run is not working with python commands (2 places)	Need to shlex.split(run_command). I have the change on my end.
> 3	T0: RF Service Validator	Coding	[========================================] 1 in 0.1s Exception Occurred: b'Traceback (most recent call last):\n  File "/home/afsanac/GitRepos/ocp-diag-ctam/ctam/RedfishServiceValidator/RedfishServiceValidator.py", line 6, in \n    from redfish_service_validator.RedfishServiceValidator import main, my_logger\n  File "/home/afsanac/GitRepos/ocp-diag-ctam/ctam/RedfishServiceValidator/redfish_service_validator/RedfishServiceValidator.py", line 12, in \n    from redfish_service_validator.metadata import getSchemaDetails\n  File "/home/afsanac/GitRepos/ocp-diag-ctam/ctam/RedfishServiceValidator/redfish_service_validator/metadata.py", line 9, in \n    from bs4 import BeautifulSoup\nImportError: No module named bs4\n'	 
> 4	Redfish validator related codes in comptool_dut.py	Design/ Framework	Can we put them in a separate file (can be existing _ifc if sounds appropriate)? My understanding is comptool_dut.py should include only DUT related functionalities	 
> 5	python command	Enhancement	Requirement on Python version. What do DMTF validators require? CTAM requires python3.10. Any possible conflict?	 
> 6	ComponentIdentifier format	Coding	As discussed in last meeting, CTAM should expect the ComponentIdentifier in PLDM metedata json file as hex format	Revert change in 2 lines converting SoftwareId ro string `str(hex(int(software_id, 16)))`
> * Could not test FW Update tests because of some system issue on my end. I reviewed the code changes and they look ok to me.
> * Since the tags are for FW update tests only at this point, I was not able to test the tag feature with the current version. In order to test, I added tags in RAS testcases (R1->L3, R2->L0, R3->L0, L1) and added `"include_tags": ["L0"], "exclude_tags": ["L1"]` in test_runner.json. Then I ran CTAM without any arguments (except workspace).
>   - Expected behavior: Only R2 will be run.
>   - Actual behavior: Only R2 was run.
>   - So the tag feature seems to work!
> 
> @MeritedHobbit - Please take a look at the table above with my concerns. We can address them either in this PR or the next one. Let me know.


1. Bug in RAS test R3: We have not fully developed this test case yet.
2.  2 & 3 are related. Will resolve by adding shlex.split
3. Redfish validator related codes in comptool_dut.py : Will create a new util file named ctam_util and move this code to the util file.
4. python command: DMTF does not depend on python version
5. ComponentIdentifier format: Will revert back this change to hex.

**Comment by MeritedHobbit** [comment]
image

Framework Changes:

1. Repo clone, use, delete moved to a new file - ctam_utils. Impacts service validator
2. Uses httppushuritargets only if it is supported in redfish_uri_config file
3. Component identifier should be a hex value in metadata json file 
4. Changes in score and result generation - Includes support for
- Compliance level based reporting & scoring
- Custom compliance weight based scoring
- Modified report look
5. test_runner.json reference changes to support the above features. "Active test suite" can now be a collection of test suites too 

Test Case Changes 

1. All test cases have a compliance_level field now (tag is now redundant). 
2. All have been assigned to L0 or L1 or L2 or L3
3. Service Validator will pass if any pass counts observed

File: ctam/utils/ctam_utils.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                    s_idx = result.index("Elapsed time:")
+                    data = result[s_idx:]
+                    # print("Index: ", data)
+                    import re
+                    res = re.findall(r"pass:\s+(\d+)", data)
+                    print("REGEX: ",res)
+                    if res and res[0].isdigit() and int(res[0]) > 0:
+                        # res = int(res[0])
+                        # if res > 0:
+                        return True
```
[the remark this answers is not in the document]
Comment (v-ajitpanda) [reply ctam/utils/ctam_utils.py:107]: Yes for this we have added logic for passget. If PassGet is more than 1 than the test is pased.

File: ctam/interfaces/comptool_dut.py [hunk 2]
```diff
@@ -125,6 +127,42 @@ def __init__(
...
+    def user_name(self):
+        return self.__user_name
+    
+    @user_name.setter
+    def user_name(self, value):
+        raise Exception("Username can not be override...")
+
+    @property
+    def user_pass(self):
+        return self.__user_name
```
[the remark this answers is not in the document]
Comment (v-ajitpanda) [reply ctam/interfaces/comptool_dut.py:152]: YEs, Will change this

File: ctam/test_runner.py [hunk 3]
```diff
@@ -547,33 +561,100 @@ def get_system_details(self):
...
+        if self.weighted_scores:
+        
+            c_data = dict(sorted(self.compliance_data.items()))
+            compliance_values = c_data.values()
+            total_execution = sum([x[2].total_seconds() for x in compliance_values])
+            total_test_cases = sum([x[3] for x in compliance_values])
+            total_passed_test_cases = sum([x[4] for x in compliance_values])
+            total_weight = sum([x[5] for x in compliance_values])
+            total_score = sum([x[6] for x in compliance_values])
+            grade = round((total_score / total_weight * 100), 2)
```
[the remark this answers is not in the document]
Comment (v-ajitpanda) [reply ctam/test_runner.py]: Could you please send me the test runner json file also. I need to check some values.

I will retest this with our system.

[the remark this answers is not in the document]
Comment (v-ajitpanda) [reply ctam/test_runner.py]: Could you please retest with our new changes. We have committed code on Friday. We are unable to find the use case.

[the remark this answers is not in the document]
Comment (afsanac-nv) [reply ctam/test_runner.py]: I tested with latest changes and here is the error I'm getting now.
```
(venv) afsanac@afsanac-dev-vm:~/ocp-diag-ctam$ python3 ctam/ctam.py -w workspace_GH/ -t H5
WorkSpace : workspace_GH/
Test Run Failed: {
    "ctam.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/ctam.py",
        "lineno": 158,
        "name": "main"
    },
    "test_runner.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/test_runner.py",
        "lineno": 178,
        "name": "_get_test_runner_config"
    },
    "type": "KeyError",
    "message": "'normalized_score'"
}

Test exited with status code*: FAIL - Test failed due to exception: 'normalized_score'
Log Directory: None

*Note: Return/Status Codes - PASS(0): All tests passed, FAIL(1): Execution/runtime failure or test failure
```
Here is the test_runner.json:
```
{
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "$id": "Need to add one",
    "title": "Test Runner Configuration",
    "description": "Configure the TestRunner obj, list of test cases overrides a test suite",
    "output_override_directory": "",
    "debug_mode": true,
    "console_log": true,
    "progress_bar":  false,
    "include_tags": [],
    "exclude_tags": [],
    "test_sequence" : [],
    "group_sequence" : [],
    "weighted_score": {
        "L0": 100, "L1": 50, "L2": 20, "L3": 0, "Others": 10
    },
    "active_test_suite": [],
    "dev_test_suite": [],
    "full_compliance_test_suite": [],
    "regression_test_suite_0": ["F0","F1", "F63", "F64", "F8", "F88", "T0"],
    "regression_test_suite_1": ["F0", "F1", "F23", "F26", "F28", "F27", "F22", "F25", "F18", "F32"],
    "regression_test_suite_2": ["F0", "F1", "F16", "F19", "F55", "F62", "F89"],
    "test_uri_response_excel":"Excel file name for checking the response"
}
```

**Comment by afsanac-nv** [comment]
Approving with some review items for future PR:

1. Total Normalized Score in Compliance Level Normalized Weighted Report table may confuse users. We may not even need that total value.

image

2. What's the significance of Grade reported in Compliance Level Normalized Weight Overall Report? I see the formula used currently is `normalized_grade = round((total_test_cases_passed / total_test_cases_available * 100), 2)`.

3. I noticed when the Ip/credentials provided in dut_config or netrc files are wrong, the error message does not really say what went wrong. This is a very common user input error, so we may want to have a clear error message. Below is an example when IP address in dut_config.yaml was incorrect.
```
WorkSpace : workspace_GH/
Group tags  []
Output Dir is :  
Test Run Failed: {
    "ctam.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/ctam.py",
        "lineno": 214,
        "name": "main"
    },
    "test_runner.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/test_runner.py",
        "lineno": 571,
        "name": "_run_group_test_cases"
    },
    "basic_health_check_test_group.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/tests/health_check/basic_health_check_group/basic_health_check_test_group.py",
        "lineno": 58,
        "name": "teardown"
    },
    "test_group.py": {
        "filename": "/home/afsanac/ocp-diag-ctam/ctam/tests/test_group.py",
        "lineno": 79,
        "name": "test_run"
    },
    "type": "NotImplementedError",
    "message": "need to call TestGroup.SetUpAssociations"
}

Test exited with status code*: FAIL - Test failed due to exception: need to call TestGroup.SetUpAssociations
Log Directory: None

*Note: Return/Status Codes - PASS(0): All tests passed, FAIL(1): Execution/runtime failure or test failure

```

**Review by afsanac-nv — APPROVED** [review]

**Pull request merged**